module/zfs/dbuf.c

```diff
@@ -2239,6 +2240,31 @@ dbuf_dirty(dmu_buf_impl_t *db, dmu_tx_t *tx)
 	return (dr);
 }
 
+static void
+dbuf_undirty_bonus(dbuf_dirty_record_t *dr)
+{
+	dmu_buf_impl_t *db = dr->dr_dbuf;
+	int max_bonuslen;
+	struct dnode *dn;
+
+	if (dr->dt.dl.dr_data != db->db.db_data) {
+		dn = DB_DNODE(db);
+		max_bonuslen = DN_SLOTS_TO_BONUSLEN(dn->dn_num_slots);
+		kmem_free(dr->dt.dl.dr_data, max_bonuslen);
+		arc_space_return(max_bonuslen, ARC_SPACE_BONUS);
+	}
+	db->db_data_pending = NULL;
+	dr = list_tail(&db->db_dirty_records);
+	list_remove(&db->db_dirty_records, dr);
```

Comment:
Why are we switching from manipulating the passed-in `dr` to the last (earliest) dirty record?  If these are the same, why do the `list_tail()`?